Add 'Securing Kubernetes Ingress with Cert-Manager' example

philrud opened the pull request:
Demonstrates how to get K8s Ingress + SDS + bundled Cert-Manager + "Let's Encrypt" setup working end to end, which is one of common (yet poorly documented) use cases that is now fully supported by Istio 1.1.

This has been tested against very recent daily builds and RCs of 1.1 and reviewed by several core contributors.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio.io/pull/3498#" title="Author self-approved">philrud</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **sdake**

If they are not already assigned, you can assign the PR to them by writing `/assign @sdake` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[content/OWNERS](https://github.com/istio/istio.io/blob/master/content/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["sdake"]} -->

istio-testing commented:
Hi @philrud. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

geeknoid commented:
Deploy preview for *preliminary-istio* failed.

Built with commit 604eeda45f9e108797e66ff7d455b0a9573e12cb

https://app.netlify.com/sites/preliminary-istio/deploys/5c7d9da2e2380f0008f8688a

philrud commented:
/assign @ozevren

istio-testing commented:
@philrud: GitHub didn't allow me to assign the following users: ozevren.

Note that only [istio members](https://github.com/orgs/istio/people) and repo collaborators can be assigned. 
For more information please see [the contributor guide](https://git.k8s.io/community/contributors/guide/#issue-assignment-in-github)

<details>

In response to [this](https://github.com/istio/istio.io/pull/3498#issuecomment-469442115):

>/assign @ozevren 


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

philrud commented:
@ozevren here's a PR for the K8S Ingress tutorial that we discussed last week, could you please help with further steps? thanks!

ozevren commented:
It looks good to me, but I'd suggest you get more reviews from people that are more familiar with the scenario.

philrud commented:
@ozevren this has been previously reviewed by @costinm, @duderino and other participants of the conversation last week (I'm simply moving the contents from personal wiki). Would you recommend additional reviewers?

ozevren commented:
/lgtm

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio.io/pull/3498#issuecomment-469791332" title="LGTM">ozevren</a>*, *<a href="https://github.com/istio/istio.io/pull/3498#" title="Author self-approved">philrud</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **sdake**

If they are not already assigned, you can assign the PR to them by writing `/assign @sdake` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[content/OWNERS](https://github.com/istio/istio.io/blob/master/content/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["sdake"]} -->

sdake commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+weight: 70
+keywords: [traffic-management,ingress,https,cert-manager,acme,sds]
+---
+
```
Need a intro `##`

philrud replied:
Looks like all other examples start with intro test right away, without any heading -- kept consistent with the rest (let me know if I misunderstood your comment).

geeknoid replied:
Yeah, no need for `#intro` section.

sdake commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+keywords: [traffic-management,ingress,https,cert-manager,acme,sds]
+---
+
+This tutorial demonstrates the use of Istio as a secure Kubernetes Ingress controller with TLS certificates issued by "Let's Encrypt". We will start with a clean Istio installation, create an example service, expose it using the Kubernetes `Ingress` resource and get it secured by instructing `cert-manager` (bundled with Istio) to manage issuance and renewal of TLS certificates that will be furthe ...
```
is "Kubernetes `INgress` right?  We just deprecated this.

philrud replied:
Yes, it is. Looks like the current consensus is that we are keeping K8s Ingress and dropping the legacy "ingress" from the installer (I'm working on a PR). This example demonstates k8s ingress support via the ingressgateway, i.e. legacy "ingress" is not required and we can definitely clean that up.

sdake commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+
+## Before you begin
+
+Install Istio by following the instructions corresponding to your platform [installation guide](/docs/setup/). Make sure to enable Ingress Gateway with Kubernetes Ingress support, SDS and [cert-manager](https://docs.cert-manager.io/) optional dependency during installation. Here's an example of how to do this for the [`helm template`](/docs/setup/kubernetes/install/helm/#option-1-install-with-hel ...
```
I think this is `(Instlalation Guide)[/docs/setuep] but could be mistaken.

`helm template`  should not be in ` as it is in a link.

philrud replied:
Copied this exact link from other examples (i.e. bookinfo) -- exact same formatting.

philrud replied:
...and removed backticks around 'helm template'

sdake commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+$ kubectl -n istio-system get service istio-ingressgateway
+{{< /text >}}
+
+Configure your DNS zone so that the domain you'd like to use for this tutorial is resolving to the external IP address of `istio-ingressgateway` service (that you've captured in the previous step). You will need a real domain name for this tutorial in order to get a TLS certificate issued. Let's store the configured domain name into an environment variable for further use:
```
can remove the `()` around captured in previous step.

philrud replied:
removed parenthesis as recommended

sdake commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+$ INGRESS_DOMAIN=mysubdomain.mydomain.edu
+{{< /text >}}
+
+Our Istio installation contains an automatically generated `Gateway` resource configured to serve the routes defined by the Kubernetes `Ingress` resources. By default it does not use SDS (Secrets Discovery Service), so we need to modify it in order to enable the delivery of the TLS certificates to the `istio-ingressgateway` via SDS:
```
Our/Your

philrud replied:
fixed

sdake commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+Our Istio installation contains an automatically generated `Gateway` resource configured to serve the routes defined by the Kubernetes `Ingress` resources. By default it does not use SDS (Secrets Discovery Service), so we need to modify it in order to enable the delivery of the TLS certificates to the `istio-ingressgateway` via SDS:
+
+{{< text bash >}}
+$ kubectl -n istio-system edit gateway
```
is this the only way to do this?   We shouldn't recommend `kubectl edit...`

philrud replied:
replaced with patch

linsun replied:
is there a helm config for this?

philrud replied:
no, see below my comment on the 'patch'

sdake commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+
+## Setting up a demo application
+
+We will be using a simple `helloworld` application for this tutorial. The following command will spin up the `Deployment` and `Service` for the demo application and expose the service using an `Ingress` resource that will be handled by `istio-ingressgateway` (notice the use of `INGRESS_DOMAIN` variable we defined earlier):
```
put the notice in an emoji please.

philrud replied:
I assume you mean a "tip" tag? Fixed. Let me know if I misunderstood this.

istio-testing commented:
New changes are detected. LGTM label has been removed.

philrud commented:
@sdake thanks for the review! addressed the issues with a second commit
This has been extensively tested by me, also @ozevren confirmed that he was able to follow the example successfully. 
I'm planning to do another round later today.

geeknoid commented:
Made a P0 per @duderino

linsun commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+  --set global.k8sIngress.gatewayName=ingressgateway \
+  --set certmanager.enabled=true \
+  --set certmanager.email=mailbox@donotuseexample.com \
+  > $HOME/istio-fetch/istio.yaml
```
suggest remove some of the defaults, e.g. ``` --set gateways.enabled=true \
   --set gateways.istio-ingressgateway.enabled=true \``` as it is not necessary.

philrud replied:
removed the defaults as recommended

linsun commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+
+{{< text bash >}}
+$ kubectl -n istio-system get service istio-ingressgateway
+{{< /text >}}
```
can we put a note around node-port?  is node-port not supported in this scenario when user exposes the ingress gateway as node port.

philrud replied:
NodePort is supported with no additional steps/changes. Added a "tip" with a corresponding installation option that user might want to consider.

linsun commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+  patch gateway istio-autogenerated-k8s-ingress --type=json \
+  -p='[{"op": "replace", "path": "/spec/servers/1/tls", "value": {"credentialName": "ingress-cert-staging", "mode": "SIMPLE", "privateKey": "sds", "serverCertificate": "sds"}}]'
+{{< /text >}}
+
```
seems we should expose this as helm and config it earlier on.

philrud replied:
Yes, we should just put this into k8s gateway generation helm under an "if gateway SDS enabled" condition. But I'm not sure if such change is worth pulling in after the 1.1 freeze. Also looks like our approach is shifting towards cleaning up installer and moving as much as possible into the docs/examples (that's what we're doing with add-ons gateways at least). I don't have strong opinion on this.

linsun commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+            backend:
+              serviceName: helloworld
+              servicePort: 5000
+---
```
aren't we supposed to configure TLS here to access the host?  I'm a little surprised it is only http here.

philrud replied:
Ingress support is not feature complete. TLS configuration is only supported via the Gateway that handles Ingress resources (that's what we're patching earlier). We have been asked about this feature, but I'm not sure what is the priority atm.

linsun commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+{{< text bash >}}
+$ curl --insecure http://$INGRESS_DOMAIN/hello
+Hello version: v1, instance: helloworld-5d498979b6-jp2mf
+{{< /text >}}
```
example with https?

philrud replied:
Should be "https" of course (and it has been tested that way), nicely spotted, thanks!

linsun commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,174 @@
+weight: 70
+keywords: [traffic-management,ingress,https,cert-manager,acme,sds]
+---
+
```
One thing that is missing for me in the tutorial is why I should choose this support over an nginx based ingress controller if my cloud provider k8s already provides that.

What would be the advantages?  It doesn't seem or make sense that I can use virtual service on the ingress so why would anyone use this over nginx based ingress controller if their cloud provider already provides one within their hosted k8s env. 

Maybe telemetry is one advantage?

philrud replied:
Elaborated on this in the intro a little bit, but could definitely use some help of tech writer here later on.

philrud commented:
Addressed @linsun 's comments, thanks for review!

philrud commented:
@linsun @geeknoid @sdake @frankbu @rcaballeromx folks, is there anything else I can do to get this merged? thanks!

geeknoid commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,178 @@
+
+This tutorial demonstrates the use of Istio as a secure Kubernetes Ingress controller with TLS certificates issued by "Let's Encrypt". While more powerful Istio concepts such as `Gateway` and `VirtualService` should be used for advanced traffic management, optional support of the Kubernetes Ingress is also available and can be used to simplify integration of legacy and third-party solutions into  ...
+
+We will start with a clean Istio installation, create an example service, expose it using the Kubernetes `Ingress` resource and get it secured by instructing `cert-manager` (bundled with Istio) to manage issuance and renewal of TLS certificates that will be further delivered to the Istio Ingress Gateway and hot-swapped as necessary via the means of SDS (Secrets Discovery Service).
```
In general, we don't use "we", use "You" instead.
Ingress Gateway -> ingress gateway
Add link for SDS

geeknoid commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,178 @@
+
+## Before you begin
+
+Install Istio by following the instructions corresponding to your platform [installation guide](/docs/setup/). Make sure to enable Ingress Gateway with Kubernetes Ingress support, SDS and [cert-manager](https://docs.cert-manager.io/) optional dependency during installation. Here's an example of how to do this for the [helm template](/docs/setup/kubernetes/install/helm/#option-1-install-with-helm- ...
```
Simply say:

[Install Istio](/docs/setup/kubernetes) making sure to enable

geeknoid commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,178 @@
+{{< /text >}}
+
+{{< tip >}}
+By default `istio-ingressgateway` will be exposed as a `LoadBalancer` service type. You may want to change that by setting `gateways.istio-ingressgateway.type` installation option to `NodePort` if this is more applicable to your Kubernetes environment.
```
setting -> setting the

geeknoid commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,178 @@
+$ kubectl -n istio-system get service istio-ingressgateway
+{{< /text >}}
+
+Configure your DNS zone so that the domain you'd like to use for this tutorial is resolving to the external IP address of `istio-ingressgateway` service that you've captured in the previous step. You will need a real domain name for this tutorial in order to get a TLS certificate issued. Let's store the configured domain name into an environment variable for further use:
```
here and elsewhere: tutorial -> example

geeknoid commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,178 @@
+{{< /text >}}
+
+{{< tip >}}
+Notice the use of `INGRESS_DOMAIN` variable we defined earlier
```
use of the

geeknoid commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,178 @@
+
+## Getting a "Let's Encrypt" certificate issued using the "cert-manager"
+
+At this point our Istio installation should have "cert-manager" up and running with two `ClusterIssuer` resources configured (for production and staging ACME-endpoints provided by "Let's Encrypt"). We will be using staging endpoint for this tutorial (feel free to try swapping `letsencrypt-staging` for `letsencrypt` to get a browser-trusted certificate issued).
```
our -> your

geeknoid commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,178 @@
+
+At this point our Istio installation should have "cert-manager" up and running with two `ClusterIssuer` resources configured (for production and staging ACME-endpoints provided by "Let's Encrypt"). We will be using staging endpoint for this tutorial (feel free to try swapping `letsencrypt-staging` for `letsencrypt` to get a browser-trusted certificate issued).
+
+In order to have a certificate issued and managed by the "cert-manager" we need to create a `Certificate` resource:
```
No double-quotes around cert-manager. It's a proper name, so no "the" in front of it either. Please review other uses in the doc accordingly.

geeknoid commented on content/docs/examples/advanced-gateways/ingress-certmgr/index.md:
```diff
@@ -0,0 +1,178 @@
+Hello version: v1, instance: helloworld-5d498979b6-jp2mf
+{{< /text >}}
+
+Note that we have to use `--insecure` flag as certificates issued by the "staging" ACME-endpoints aren't trusted.
```
to use the

philrud commented:
Thanks, @geeknoid! Made edits as recommended -- see the most recent commit. Let me know if things are looking okay and I'll squash it.
CircleCI build is failing due to unrelated reason (I don't seem to have necessary permissions to retrigger it).

geeknoid commented:
Please rebase and repush, that should fix Ci.

philrud commented:
done, CircleCI keeps failing though

geeknoid commented:
If CI keeps failing, I'll take a look when I'm free later today. I can merge then.

In the meantime, can you try "make lint" and see what happens?

philrud commented:
looks like third time's the charm -- it's green now

geeknoid commented:
/lgtm
/approve

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio.io/pull/3498#issuecomment-471038588" title="Approved">geeknoid</a>*, *<a href="https://github.com/istio/istio.io/pull/3498#issuecomment-469791332" title="LGTM">ozevren</a>*, *<a href="https://github.com/istio/istio.io/pull/3498#" title="Author self-approved">philrud</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[content/OWNERS](https://github.com/istio/istio.io/blob/master/content/OWNERS)~~ [geeknoid]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
